# IMPORTING LIBRARIES 

In [122]:
import pandas as pd
import numpy as np
import os 

# IMPORTING DATA

# creating main folder path

In [123]:
path = r'C:\Users\João Pedro\MAR-Instacart Basket Analysis'

In [124]:
path

'C:\\Users\\João Pedro\\MAR-Instacart Basket Analysis'

# importing dataset products.csv

In [125]:
df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'Original Data' , 'products.csv') , index_col = False)

# importing dataset orders_wrangled.csv

In [126]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data' , 'orders_wrangled.csv') , index_col = False)

# DATA CONSISTENCY CHECKS

In [127]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# MIXED TYPE DATA

# create a dataframe

In [128]:
df_test = pd.DataFrame()

# create a mixed type column

In [129]:
df_test['mix'] = ['a','b', 1 , True]

In [130]:
df_test.head()

mix
0     a
1     b
2     1
3  True

# check for mixed types

In [131]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


# fix mixed_type column

In [132]:
df_test['mix'] = df_test['mix'].astype('str')

# MISSING VALUES

# obtain the total number of missing observations

In [133]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

# create a subset containing only missing values

In [134]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [135]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [136]:
df_prods.shape

(49693, 5)

# create a subset containing only non-missing values

In [137]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [138]:
df_prods_clean.shape

(49677, 5)

# DUPLICATE VALUES

# check for full duplicates within the df_prods_clean\

In [139]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [140]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [141]:
df_prods_clean.shape

(49677, 5)

# Dropping the duplicate values by creating a new dataframe

In [142]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [143]:
df_prods_clean_no_dups.shape

(49672, 5)

# TASK 4.5

# 2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [144]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# The 'days_since_prior_order' column contains numerous missing values, evident from its lower count compared to the other columns. #

# The 'days_since_prior_order' column exhibits a minimum value of 0 and a maximum of 30, with a median (Q2) of 7. No discrepancies were detected. #

# The 'order_number' column ranges from a minimum value of 1 to a maximum of 100, with a median (Q2) of 11. No discrepancies were identified. #

# 3 Check for mixed-type data in your df_ords dataframe

In [145]:
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)
else:
    print('None of the columns have mix data')

None of the columns have mix data


# No mixed_type data was found # 

# 4  If you find mixed-type data, fix it. The column in question should contain observations of a single data type

# 5 Run a check for missing values in your df_ords dataframe

In [146]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

# The 'days_since_prior_order' column contains 206,209 missing values, likely indicating first orders made by customers. Consequently, any initial order will result in a missing value in the 'days_since_prior_order' column #

# 6 Address the missing values using an appropriate method

# Retaining the missing values is essential to preserve crucial information about customers making their first order. Consequently, here is my plan to address the issue : #

# 1. Verifying that the frequency of missing values aligns with the frequency of order_number=1. #

# 2. Generating a new column to indicate whether orders are new (True) or not (False) #

# Create a subset containing only missing values

In [147]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [148]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

# Check the frequency of order_number=1 in df_nan

In [149]:
df_nan['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

# Check the frequency of order_number=1 in df:ords

In [150]:
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

# The occurrence of missing values (206,209) corresponds to the frequency of order_number=1 #

# # Create a function to determine if the order in question is a first order or not #

In [151]:
def is_first_order(row):
    if row['order_number'] == 1:
        return True
    else:
        return False

# Apply the is_first_order function to every row of the df_ords dataframe

In [152]:
df_ords.apply(lambda row: is_first_order(row), axis=1)

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

# Create a new column in df_ords to flag the reason for missing values\

In [153]:
df_ords['first_order']=df_ords.apply(lambda row: is_first_order(row) ,axis=1)

In [154]:
df_ords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           1   2398795        1    prior             2                   3   
2           2    473747        1    prior             3                   3   
3           3   2254736        1    prior             4                   4   
4           4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

# Final check for the frequency of first_order=True

In [155]:
df_ords['first_order'].value_counts()

first_order
False    3214874
True      206209
Name: count, dtype: int64

# Missing values handled correctly #

# 7 Run a check for duplicate values in your df_ords data

In [156]:
df_ords_clean = df_ords

In [157]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [158]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

# 8 Address the duplicates using an appropriate method

In [160]:
df_ords_clean.shape

(3421083, 9)

# 9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.